File: Php/tests/AppTests/User/PasswordSenderTests.php
namespace tests\AppTests\User{

	use app\User\ForgotPassword\PasswordSenderImpl;
	use Domain\Classes\User;
	use tests\SharbatUnitTestCase;
	use app\Urls\AuthLoginUrlModel;
	use core\ClassNames;
	use tests\TestHelper;

	class PasswordSenderTests extends SharbatUnitTestCase {

		/** @var PasswordSenderImpl */
		private $passwordSender;
		private $messageSenderFactory;
		private $settings;

		function setUp(){
			parent::setUp();
			messageSenderFactory = registerFake(ClassNames::$MessageSenderFactory);
			settings = registerFake(ClassNames::$Settings);
			passwordSender = getInstance('app\User\ForgotPassword\PasswordSenderImpl');
		}

		function test_GetSubject_ReturnsSubjectThatIncludesHomegameDisplayName(){
			$subject = passwordSender.getSubject();

			assertTrue(strlen($subject) > 0);
		}

		function test_GetBody_ReturnsBodyThatIncludesLoginUrl(){
			$loginUrl = new AuthLoginUrlModel();
			$body = passwordSender.getBody('anypassword');

			assertTrue(strpos($body, $loginUrl.url));
		}

		function test_GetBody_ReturnsBodyThatIncludesPassword(){
			$password = 'generated-password';
			$body = passwordSender.getBody($password);

			assertTrue(strpos($body, $password));
		}

		function test_Send_CallsSendOnMessageSender(){
			$user = new User();
			$messageSender = TestHelper::getFake('integration\Message\MessageSender');
			messageSenderFactory.returns("getMessageSender", $messageSender);
			$messageSender.expectOnce("send");

			passwordSender.send($user, "anypassword");
		}

	}

}

File: Tests.Application/Services/PasswordSenderTests.cs
using Application.Services;
using NUnit.Framework;
using Tests.Common;
using Tests.Common.FakeClasses;

namespace Tests.Application.Services
{
    public class PasswordSenderTests : MockContainer
    {
        [Test]
        public void Send_SendsMessageWithCorrectSubjectBody()
        {
            const string email = "a";
            const string subject = "b";
            const string body = "c";
            const string password = "d";
            var user = new UserInTest(email: email);
            GetMock<IPasswordMessageBuilder>().Setup(o => o.GetSubject()).Returns(subject);
            GetMock<IPasswordMessageBuilder>().Setup(o => o.GetBody(password)).Returns(body);

            var sut = GetSut();
            sut.Send(user, password);

            GetMock<IMessageSender>().Verify(o => o.Send(email, subject, body));
        }

        private PasswordSender GetSut()
        {
            return new PasswordSender(
                GetMock<IMessageSender>().Object,
                GetMock<IPasswordMessageBuilder>().Object);
        }
    }
}
